xiilib/_gunicorn/charm.py

```diff
@@ -0,0 +1,191 @@
+        self.framework.observe(
+            self.on.secret_storage_relation_changed,
+            self._on_secret_storage_relation_changed,
+        )
+        self.framework.observe(self.on.update_status, self._on_update_status)
+        for database, database_requirer in self._database_requirers.items():
+            self.framework.observe(
+                database_requirer.on.database_created,
+                getattr(self, f"_on_{database}_database_database_created"),
+            )
+            self.framework.observe(
+                database_requirer.on.endpoints_changed,
+                getattr(self, f"_on_{database}_database_endpoints_changed"),
+            )
+            self.framework.observe(
+                self.on[database_requirer.relation_name].relation_broken,
+                getattr(self, f"_on_{database}_database_relation_broken"),
+            )
+
+    def _on_config_changed(self, _event: ops.EventBase) -> None:
+        """Configure the flask pebble service layer.
+
+        Args:
+            _event: the config-changed event that triggers this callback function.
+        """
+        self.restart()
+
+    def _on_rotate_secret_key_action(self, event: ops.ActionEvent) -> None:
+        """Handle the rotate-secret-key action.
+
+        Args:
+            event: the action event that trigger this callback.
+        """
+        if not self.unit.is_leader():
+            event.fail("only leader unit can rotate secret key")
+            return
+        if not self._secret_storage.is_initialized:
+            event.fail("flask charm is still initializing")
+            return
+        self._secret_storage.reset_secret_key()
+        event.set_results({"status": "success"})
+        self.restart()
+
+    def _on_secret_storage_relation_changed(self, _event: ops.RelationEvent) -> None:
+        """Handle the secret-storage-relation-changed event.
+
+        Args:
+            _event: the action event that triggers this callback.
+        """
+        self.restart()
+
+    def _update_app_and_unit_status(self, status: ops.StatusBase) -> None:
+        """Update the application and unit status.
+
+        Args:
+            status: the desired application and unit status.
+        """
+        self.unit.status = status
+        if self.unit.is_leader():
+            self.app.status = status
+
+    def restart(self) -> None:
+        """Restart or start the flask service if not started with the latest configuration."""
+        try:
+            self._wsgi_app.restart()
+        except CharmConfigInvalidError as exc:
+            self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
+            return
+        self._update_app_and_unit_status(ops.ActiveStatus())
+
+    def _on_update_status(self, _: ops.HookEvent) -> None:
+        """Handle the update-status event."""
+        if self._database_migration.get_status() == DatabaseMigrationStatus.FAILED:
+            self.restart()
```

Comment:
At some point I think we'll be able to use pebble notices for this to allow feedback directly from the migration script to our charm

Reply:
I think that's only available for juju >= 3.4?

Reply:
Yeap, it's a "for later" comment